### PR TITLE
datadog: Config (final) portion of new tracing library

### PR DESCRIPTION
**Pull request by dgoffredo**
This is the final part of the work proposed in https://github.com/envoyproxy/envoy/pull/23958.

~~This PR additionally contains all of the changes from <https://github.com/envoyproxy/envoy/pull/26148>. This diff will shrink when that PR is merged.~~

This revision alters the `DatadogTracerFactory` in `source/extensions/tracers/datadog/config.h` to produce instances of the new dd-trace-cpp based implementation of the Datadog tracer, instead of the old dd-opentracing-cpp based one.

This revision actually alters the implementation of the Datadog tracer. Previous pull requests have only added code for use here.

This revision also moves some unit tests around:

- Two tests that were part of `datadog_tracer_impl_test.cc` have been moved into `agent_http_client_test.cc`, because they're about HTTP requests.
- One test that was part of `datadog_tracer_impl_test.cc` has been removed because its behavior is already covered by tests added in other PRs in this series.
- The remaining tests in `datadog_tracer_impl_test_.cc` were moved into `config_test.cc`, since they all now have to do with the configuration of the tracer and the resulting behavior.

Finally, all references to dd-opentracing-cpp and msgpack have been removed from the repo, and the documentation (READMEs) and `demo/` directory from https://github.com/envoyproxy/envoy/pull/23958 have been added.

### CODE REVIEW

**Comment by alyssawilk** [comment]
/wait on CI - can you check if clang and coverage are real?

**Comment by dgoffredo** [comment]
Coverage is 100%
https://storage.googleapis.com/envoy-pr/e3829ca/coverage/source/extensions/tracers/datadog/index.html

It looks like the fact that Datadog no longer uses OpenTracing has reduced the test coverage for the OpenTracing code:
https://dev.azure.com/cncf/envoy/_build/results?buildId=131922&view=logs&jobId=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&j=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&t=00ade0d6-ffae-5af0-b5e0-e33b1813c8bc

There was what looks like an unrelated, spurious failure here:
https://dev.azure.com/cncf/envoy/_build/results?buildId=131922&view=logs&jobId=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&j=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&t=c7ff987c-56c2-5ec2-60f4-6c5e34f756c3

Clang did produce some warnings:
https://dev.azure.com/cncf/envoy/_build/results?buildId=131922&view=logs&jobId=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=9690e737-a3af-5ce5-a739-e50023b7ccb0

It looks like I have `using` directives that are not needed. Will fix that.

What should we do about the OpenTracing code coverage?

**Comment by dgoffredo** [comment]
Nope, I can't remove the OpenTracing code. Doing so would disable the following feature: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/dynamic_ot.proto

Until explicit support for OpenTracing is removed (might be good to continue support, IMO), we'll still have the `extentions/tracers/common/ot` library, whose implementation how has inadequate test coverage due to the change in one of its former clients (`extensions/tracers/datadog`) here.

**Comment by moderation** [comment]
/lgtm deps

Awesome to seeing this about to land

**Comment by dgoffredo** [comment]
This PR is now up-to-date with `main` (which includes a PR that was previously part of this diff).

Here's a recap of what's in this PR:

- The "config" class that's registered as a tracing extension. It's a factory that takes proto and produces tracer.
- What was the "impl" class and its tests has been broken up into parts: some in config, some in agent http client.
- Documentation was added in the form of README files.
- A demo setup was added in its own directory.
- The Datadog tracing extension now uses the new dd-trace-cpp based code, instead of the previous dd-opentracing-cpp based code.
- The dd-opentracing-cpp external library dependency has been removed.
- The msgpack-c external library dependency has been removed.

One thing I'll want to do before this is merged is update the version of dd-trace-cpp again. The hard-coded version number in the source is wrong in the latest release. Will fix that ASAP.

I commented above about how the coverage CI check is failing in the shared OpenTracing code. I think this is because the Datadog tracer tests were covering some of the underlying middleware code, but now they're not. Not sure what to do about that.

**Comment by dgoffredo** [comment]
@cgilmour Please do a review of this PR for Datadog-specific concerns. In particular, I modified the testing of what used to be the one "_impl..." test, which I believe you wrote. Want to make sure I didn't leave out anything important.

**Comment by cgilmour** [comment]
I've had a look through, and it seems that test coverage for this is a superset of what was in place before.
The original implementation "shared" some tests from `tracers/common/ot` and provided some datadog-specific ones in its own section of the overall implementation.

Now these are all covered by `tracers/datadog` along with additional ones that are focused on the new implementation.

It'd be good to have a specific test around the "healthcheck" scenario, where a trace starts off with a decision to sample, then explicitly gets told not to because it's a healthcheck, and the outcome is that it's not sampled. This might be there implicitly from other tests, and I don't expect adding it to surface a bug at all. It's really to just suggested to lock-in the behavior.

**Comment by phlax** [comment]
just flagging you can do this on PRs

/coverage

**Comment by dgoffredo** [comment]
> It'd be good to have a specific test around the "healthcheck" scenario, where a trace starts off with a decision to sample, then explicitly gets told not to because it's a healthcheck, and the outcome is that it's not sampled. This might be there implicitly from other tests, and I don't expect adding it to surface a bug at all. It's really to just suggested to lock-in the behavior.

As far as I can see, healthcheck-ness enters into control here:

- https://github.com/envoyproxy/envoy/blob/ee66b11548a5e65ca1e4e59cd0fb8fc743bed654/source/common/http/conn_manager_impl.cc#L1265-L1271
- https://github.com/envoyproxy/envoy/blob/ee66b11548a5e65ca1e4e59cd0fb8fc743bed654/source/common/tracing/tracer_impl.cc#L70-L74
- https://github.com/envoyproxy/envoy/blob/ee66b11548a5e65ca1e4e59cd0fb8fc743bed654/source/extensions/tracers/datadog/tracer.cc#L119-L122

What matters there is that `traced == false`, which is covered by <https://github.com/envoyproxy/envoy/blob/ee66b11548a5e65ca1e4e59cd0fb8fc743bed654/test/extensions/tracers/datadog/tracer_test.cc#L92-L96>.

Ah, but that test doesn't guarantee that the tracer would otherwise keep the trace. That's a change I can make.

**Comment by jmarantz** [comment]
Once @cgilmour approves I'll do a pass. Thanks!

File: source/extensions/tracers/datadog/demo/http.js
Comment (jmarantz) [inline source/extensions/tracers/datadog/demo/http.js:15]: nit: any particular reason you are sometimes using `function` in this file and sometimes using `=>` ?

File: test/extensions/tracers/datadog/agent_http_client_test.cc
Comment (jmarantz) [inline test/extensions/tracers/datadog/agent_http_client_test.cc:31]: nit: usually we put these using decls at top level rather than repeating at call-sites.

File: test/extensions/tracers/datadog/agent_http_client_test.cc
Comment (jmarantz) [inline test/extensions/tracers/datadog/agent_http_client_test.cc:522]: Are these braced stanzas independent or in sequence?

If they are independent consider putting the 2 init lines in a helper function in the test fixture, and making each of these braced stanzas each be separate TEST_F methods.

**Review by jmarantz — COMMENTED** [review]
looks great! just a few nits

File: source/extensions/tracers/datadog/demo/http.js
Comment (dgoffredo) [reply source/extensions/tracers/datadog/demo/http.js:15]: I'm inconsistent, especially when the code began in another of my projects and then got extended to purpose here.

There are a few real differences between arrow functions and `function` in javascript, but mostly I use arrow functions for the short ones and `function` for the longer ones. Except when I'm inconsistent.

File: test/extensions/tracers/datadog/agent_http_client_test.cc
Comment (dgoffredo) [reply test/extensions/tracers/datadog/agent_http_client_test.cc:522]: They're in sequence due to the accumulation of values in `stats_`.

This test is moved from https://github.com/envoyproxy/envoy/blob/7a1b9b5412420c2cd18c2d6a0c115eac0a94ed71/test/extensions/tracers/datadog/datadog_tracer_impl_test.cc#L265.

They could probably be separated into individual test methods with a little renovation. I'd prefer to leave it as-is, but if it smells, let me know.

**Comment by dgoffredo** [comment]
Joshua, thanks for all of the time and consideration you've given to these pull requests over the last few months.

The result we're arriving at is better than what I proposed originally, and I've learned a good deal from interacting with you, Wang, and others.

I look forward to our future PRs when the bug reports start rolling in. :)

**Review by jmarantz — DISMISSED** [review]
basically great; couple of more minor suggestions. If you find them helpful please execute otherwise I can merge.

**Comment by jmarantz** [comment]
oh and still needs another round of format fix.

**Comment by dgoffredo** [comment]
After addressing Joshua's review comments, I pulled `main` and merged it into my local branch. I then ran the code formatter via `ci/run_envoy_docker.sh './ci/do_ci.sh format'`, and it modified the following unrelated code:
```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto
	modified:   api/envoy/extensions/filters/http/jwt_authn/v3/config.proto
	modified:   bazel/external/http_parser/http_parser.h
	modified:   source/extensions/common/wasm/ext/envoy_null_plugin.h
	modified:   third_party/android/LocalArray.h
	modified:   third_party/android/ScopedFd.h
	modified:   third_party/android/ifaddrs-android.h
	modified:   tools/testdata/check_format/attribute_packed.cc
	modified:   tools/testdata/check_format/code_conventions.cc
	modified:   tools/testdata/check_format/condvar_wait_for.cc
	modified:   tools/testdata/check_format/counter_from_string.cc
	modified:   tools/testdata/check_format/gauge_from_string.cc
	modified:   tools/testdata/check_format/grpc_init.cc
	modified:   tools/testdata/check_format/grpc_shutdown.cc
	modified:   tools/testdata/check_format/long_line.cc
	modified:   tools/testdata/check_format/real_time_source.cc
	modified:   tools/testdata/check_format/real_time_system.cc
	modified:   tools/testdata/check_format/shared_mutex.cc
	modified:   tools/testdata/check_format/sleep.cc
	modified:   tools/testdata/check_format/source/raw_try.cc
	modified:   tools/testdata/check_format/std_any.cc
	modified:   tools/testdata/check_format/std_atomic_free_functions.cc
	modified:   tools/testdata/check_format/std_get_if.cc
	modified:   tools/testdata/check_format/std_holds_alternative.cc
	modified:   tools/testdata/check_format/std_make_optional.cc
	modified:   tools/testdata/check_format/std_monostate.cc
	modified:   tools/testdata/check_format/std_optional.cc
	modified:   tools/testdata/check_format/std_string_view.cc
	modified:   tools/testdata/check_format/std_variant.cc
	modified:   tools/testdata/check_format/std_visit.cc
	modified:   tools/testdata/check_format/test/register_factory.cc
	modified:   tools/testdata/check_format/test_naming.cc
	modified:   tools/testdata/check_format/throw.cc
```
As such, I won't be pushing those changes (neither the formatting nor the merge-from-`main`).

**Comment by jmarantz** [comment]
Formatting can be a bit of a mess if your local toolchain does not match what CI uses. I've managed to align my environment to CI by using `clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04`.

I also want to call your attention to a script I wrote to help do fast incremental format-fixes (for a subset of the checks). This formats only files modified since your last `git commit` (by default) or in your PR (option `-main`).

See `tools/local_fix_format.sh`

**Review by jmarantz — DISMISSED** [review]
oh nice. re-approving now as gesture of forward progress but I'll have to re-approve once CI is happy, I think.

/wait

**Comment by dgoffredo** [comment]
> Formatting can be a bit of a mess if your local toolchain does not match what CI uses. I've managed to align my environment to CI by using `clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04`.

`ci/run_envoy_docker.sh './ci/do_ci.sh format'` runs in docker. Might it not be in sync with CI's image? I'll take a look.

> See `tools/local_fix_format.sh`

Nice, thanks.

> I'll have to re-approve once CI is happy, I think.

Let's see what it doesn't like...

**Comment by jmarantz** [comment]
Docker always caused me such operational grief that I've opted for trying to keep my native tools in sync with CI. I would've thought that would insulate you.

**Comment by dgoffredo** [comment]
One issue is that this part of the CI https://dev.azure.com/cncf/envoy/_build/results?buildId=133809&view=logs&j=55bc2ab9-36c1-5ab8-58a6-8c258728eeff&t=46ad7877-9039-5f10-4756-88545769f7c8 is saying that my `demo/envoy.yaml` has a line that's too long:
```yaml
                  inline_string: "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH):256% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" \"%RESP(X-AMZN-RequestId)%\" \"%REQ(X-DATADOG-TRACE-ID)%\" \"%REQ(X-DATADOG-PARENT-ID)%\"\n"
```
However, when I run the code formatter (`ci/run_envoy_docker.sh 'ci/do_ci.sh format'`, per the top of the CI job's output), there is no change.

I'll see how I can wrap/shorten that string literal in the YAML file.

**Comment by dgoffredo** [comment]
It looks like the remaining CI failures are spurious. A timeout in iOS tests, and a bazel build server 5xx error in envoy-presubmit.

Should we poke it?

**Comment by phlax** [comment]
/retest

apologies im running release CI and i think that is breaking stuff

**Review by jmarantz — APPROVED** [review]

**Comment by jmarantz** [comment]
So if this passes the required tests I can go ahead and merge?

**Comment by dgoffredo** [comment]
> So if this passes the required tests I can go ahead and merge?

Sounds good to me.

**Comment by jmarantz** [comment]
I meant that question for @phlax that I'd merge despite failing non-required tests for ios and swift.

**Comment by phlax** [comment]
re mobile CI - probs yes if its not required then i guess we just need to ignore it for now

but i think there are some other issues

```console
Per-extension coverage failed:
Code coverage for source/extensions/tracers is lower than limit of 95.8 (95.5)
Code coverage for source/extensions/tracers/common is lower than limit of 73.8 (68.1)
Code coverage for source/extensions/tracers/common/ot is lower than limit of 71.7 (65.6)
```

https://dev.azure.com/cncf/envoy/_build/results?buildId=133953&view=logs&j=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&t=00ade0d6-ffae-5af0-b5e0-e33b1813c8bc&l=4761

**Comment by dgoffredo** [comment]
```
Per-extension coverage failed:
Code coverage for source/extensions/tracers is lower than limit of 95.8 (95.5)
Code coverage for source/extensions/tracers/common is lower than limit of 73.8 (68.1)
Code coverage for source/extensions/tracers/common/ot is lower than limit of 71.7 (65.6)
```
The Datadog-specific code that I modified used to use `[..]/tracers/common/ot`, but it no longer does. This resulted in `common/ot`'s test coverage dropping, presumably because the Datadog code used to explore some code paths in `common/ot` not covered by `common/ot`'s own tests.

I'd prefer not to write additional tests for `common/ot`. What do you think?

**Comment by jmarantz** [comment]
Was datadog the only in-repo user of .../common/ot ? It would be excellent to remove it in that case :)

**Comment by phlax** [comment]
> Was datadog the only in-repo user of .../common/ot ? It would be excellent to remove it in that case :)

cc @moderation 

i think that is something we are waiting for - i saw on one of the PRs it mentioned removing it and was wondering if we were going to do so before release

**Comment by dgoffredo** [comment]
See my comments above from a few weeks ago.

https://github.com/envoyproxy/envoy/pull/26284#issuecomment-1483034813

**Comment by jmarantz** [comment]
Got it. I think I agree that we do not need to hold this PR hostage to the testing of that library. However, can you open a bug to the author of that subsystem that it needs to self-test?

I'm OK otherwise with tamping down the coverage for that subdirectory in this PR though I'll ask for a senior maintainer sign-off for that.

**Comment by dgoffredo** [comment]
> can you open a bug to the author of that subsystem that it needs to self-test?

I'll do that.

**Comment by dgoffredo** [comment]
diff

There's hardly any difference in the code coverage. I'll look into a patch tomorrow; probably faster that way.

**Comment by jmarantz** [comment]
Sorry I had one more thought. This is up to you because it's an extension. But do you want to consider spinning out yet one more PR, that would add the new capability, but flag-controlled. The old version would still be available for some bounded amount of time before you (a) flip the default and (b) drop the old functionality?

Again this is up to you.

**Comment by moderation** [comment]
This is a Datadog call, however the existing extension is available in the 1.25.x release cycle. The new extension would be in the 1.26x release chain. Seems OK vs. a flag (and then flag maintenance)

**Comment by dgoffredo** [comment]
I prefer to release in `1.26x` without a toggle.

**Comment by dgoffredo** [comment]
> There's hardly any difference in the code coverage. I'll look into a patch tomorrow; probably faster that way.

Now that I've looked into it, it's quite hard to cover the circled code in the current unit test. Instead, I'll open an issue @'ing the `CODEOWNERS` for that test.

In light of that, could we override the coverage requirement of `[...]/common/ot` for this pull request?

**Comment by jmarantz** [comment]
Yes, in my opinion we can override the coverage requirement in this PR. In that case I think we should get a senior maintainer stamp however.

**Comment by dgoffredo** [comment]
https://github.com/envoyproxy/envoy/issues/26772

**Comment by dgoffredo** [comment]
I've reduced the required test coverage threshold per https://github.com/envoyproxy/envoy/issues/26772#issuecomment-1509787199

**Comment by dgoffredo** [comment]
/retest

"envoy-presubmit (Checks (Linux x64) Linux x64 clang_tidy)" failed due to an HTTP 500 while fetching a source package.

**Review by jmarantz — DISMISSED** [review]
/assign-from @envoyproxy/senior-maintainers

**Comment by jmarantz** [comment]
i added a senior maintainer pass mainly for the coverage override. A shared /ot component was relying on Datadog for testing coverage, and Datadog is no longer user/testing that shared component; the owner of the 'ot' component should own the fixing of its test coverage.

**Comment by cgilmour** [comment]
Just in case it comes up during the senior-maintainer review, this has my approval as a replacement for the original implementation. It's mentioned on other PRs, but adding here as well.

@dgoffredo and I are on the same team at Datadog, and appreciate this being a feature in Envoy.

**Comment by wbpcode** [comment]
friendly ping @moderation for deps check.

File: test/per_file_coverage.sh
Comment (alyssawilk) [inline test/per_file_coverage.sh:63]: >i added a senior maintainer pass mainly for the coverage override. A shared /ot component was relying on Datadog for testing coverage, and Datadog is no longer user/testing that shared component; the owner of the 'ot' component should own the fixing of its test coverage.

If this PR moves datadog off of shared code and means the shared code is no longer tested, I think adding back some tests is within scope of the PR.

**Review by alyssawilk — CHANGES_REQUESTED** [review]

File: test/per_file_coverage.sh
Comment (dgoffredo) [reply test/per_file_coverage.sh:63]: @alyssawilk precedence says otherwise: https://github.com/envoyproxy/envoy/commit/eb521f42e760b9e5cb6ca544e5c87523f0592fa9#diff-0ac04ef31c616358d8234bd473469c1a233ecf37947a13475cce764e7f6c43eeR79-R81

Comment (alyssawilk) [reply test/per_file_coverage.sh:63]: That was a removal which blocked key dependency upgrades and we were willing to make allowances.  I don't think such an exception applies here.

**Comment by jmarantz** [comment]
@wbpcode 

My personal feeling is that asking datadog to write tests for code they are no longer using isn't out of the question, but I can understand the pushback to that toil which doesn't add value to Datadog.

I chatted with Alyssa about this and I think my recommendation is to cut a separate PR to move the ot library out of extensions and into contrib, thus avoiding the coverage metric check. AFAICT there is no dependence on this other than datadog, and it is not adequately tested on its own.

Another option which also makes sense to me is to, in this PR, delete the ot directories and their files. That would be more extreme, but it would wake up anyone that cares to re-add this (and contribute tests to unblock that PR).

As @wbpcode is an owner of this extension WDYT?

@htuch may also have context on the origins of the ot libraries.

**Comment by alyssawilk** [comment]
@Shikugawa @basvanbeek the other codeowners

**Comment by dgoffredo** [comment]
I'm currently working on modifications to the `common/ot` code that might resolve this. It will require a little circumvention but will probably come out fine.

> [...] AFAICT there is no dependence on this other than datadog, and it is not adequately tested on its own.
>
> Another option which also makes sense to me is to, in this PR, delete the ot directories and their files. That would be more extreme, but it would wake up anyone that cares to re-add this (and contribute tests to unblock that PR).

`common/ot` is depended upon additionally by `dynamic_ot`: https://github.com/envoyproxy/envoy/blob/6d6661caa18007375dea0cfae65c041df40726fd/source/extensions/tracers/dynamic_ot/BUILD#L24

The relevant feature is documented here: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/dynamic_ot.proto

It allows a user to bring their own build of an OpenTracing-compatible tracer and `dlopen` it into Envoy on start.

:shrug: If you're willing to remove that feature, then the `common/ot` code can be removed also.

**Comment by wbpcode** [comment]
> As @wbpcode is an owner of this extension WDYT?

In fact, I have no strong point to this. Because I don't know if there are actual users of `common/ot` or `dynamic_ot`. I had never got related feedbacks from community, if my memory is correct.

If the community think we should keep these extensions, I can add some tests to improve the coverage in the next week and unblock this PR (I will have some free time at that time.)

**Comment by wbpcode** [comment]
> I'm currently working on modifications to the common/ot code that might resolve this. It will require a little circumvention but will probably come out fine.

If you can resolve it directly it would great.

**Comment by dgoffredo** [comment]
I pushed a commit that covers the following code: https://github.com/envoyproxy/envoy/blob/4c12d8eda176081a493968709d510065d715347b/source/extensions/tracers/common/ot/opentracing_driver_impl.cc#L68-L75

by adding some new boilerplate to the `common/ot` unit test: https://github.com/envoyproxy/envoy/pull/26284/commits/a7bf4fd69e736d1d33a2b46a4a76a103b6eae5b9

The difficulty is that the existing unit test uses `opentracing::mocktracer::MockTracer` to interactive with the driver under test. `opentracing::mocktracer::MockTracer` will not call the "for each" flavor of header extraction. It's just not in the `opentracing-cpp` code. `opentracing-cpp` is deprecated and hasn't been touched in a while, so I don't think modifying that library and cutting a release is a viable option.

Instead, I reproduce a skeleton of an `opentracing::Tracer` (together with its `opentracing::Span` and `opentracing::SpanContext`) that does nothing except call the "for each" flavor when extraction is requested.

**Comment by alyssawilk** [comment]
sweet!  No changes to core code and no changes to coverage so I'll recuse myself and let josh merge.

**Comment by dgoffredo** [comment]
@alyssawilk we are go for launch, standby for ignition sequence

**Review by jmarantz — APPROVED** [review]
@alyssawilk has deferred to me. LGTM!

**Comment by dgoffredo** [comment]
🎉🎉🎉🎉🎉🎉🎉🎉

Once again, thank you @jmarantz, @wbpcode, and everyone who contributed to the review of these 10 pull requests.

I learned a lot and am looking forward to collaborating again.

**Comment by dgoffredo** [comment]
@moderation

> This is a Datadog call, however the existing extension is available in the 1.25.x release cycle. The new extension would be in the 1.26x release chain. Seems OK vs. a flag (and then flag maintenance)

I downloaded the [zipped source code][1] from the [v1.26.1][2] release, and it does not include the changes in this PR, thus v1.26.1 contains the old extension, not the new one.

Might a subsequent release in the v1.26.x series contain these changes, or will we have to wait for v1.27.x? I'd prefer to have these changes available as soon as possible. The changes were initially proposed in November of 2022.

[1]: https://github.com/envoyproxy/envoy/archive/refs/tags/v1.26.1.zip
[2]: https://github.com/envoyproxy/envoy/releases/tag/v1.26.1

**Comment by phlax** [comment]
> Might a subsequent release in the v1.26.x series contain these changes

its possible we could backport the changes to 1.26.x, but i think that we should not as that is usually reserved for fixes

the 1.27 release is due for 2023/07/14

**Pull request merged**
